java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -646,13 +684,15 @@ public void visitVariable(VariableTree tree) {
     public void visitWhileStatement(WhileStatementTree tree) {
       Set<Symbol.VariableSymbol> assignedSymbols = extractor.findAssignedVariables(tree);
       invalidateAssignedVariables(assignedSymbols);
-      PackedStatementStates nextStates = PackedStatementStates.instantiate();
-      for (ExecutionState state : currentStates) {
-        PackedStates conditionStates = evaluateCondition(state, tree.condition());
-        PackedStatementStates loopStates = evaluateStatement(conditionStates.trueStates, tree.statement());
-        if (!conditionStates.falseStates.isEmpty() || !loopStates.isEmpty()) {
-          state.mergeRelations(Iterables.concat(conditionStates.falseStates, loopStates));
-          nextStates.addState(state);
+      PackedStatementStates nextStates = new PackedStatementStates(currentStates);
+      for (ExecutionState executionState : currentStates) {
+        for (ExecutionState conditionState : evaluateCondition(executionState, tree.condition())) {
+          SymbolicValue condition = conditionState.popLast();
+          if (condition == SymbolicValue.BOOLEAN_TRUE) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)